# Preparing the data for Clustering and EDA

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pandasql import sqldf

from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score
from sklearn.cluster import DBSCAN, KMeans

I am reading in the two excel sheets separately and will use SQL to combine them for clustering and to create a revenue file for use in Time Series analysis.

In [7]:
year0910 = pd.read_excel('../data/online_retail_II.xlsx', sheet_name=0)

In [8]:
year1011 = pd.read_excel('../data/online_retail_II.xlsx', sheet_name=1)

#### Columns

In [13]:
year0910.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

In [20]:
year0910.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom

#### Shape

In [9]:
year0910.shape

(525461, 8)

In [21]:
year0910.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object

#### Number of invoices vs items purchased

In [22]:
year0910.groupby('Invoice')['InvoiceDate'].count()

Invoice
489434      8
489435      4
489436     19
489437     23
489438     17
           ..
C538121     1
C538122     1
C538123     3
C538124     3
C538164     1
Name: InvoiceDate, Length: 28816, dtype: int64

There were 525461 different items purchased, on 28816 separate invoices.

In [10]:
year1011.shape

(541910, 8)

In [11]:
year0910.isna().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [12]:
year1011.isna().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

## Combining data and grouping by date for Time Series

In [24]:
sql = """
SELECT *
FROM year0910
UNION
SELECT *
FROM year1011
"""

df = sqldf(sql)
df.head()

Invoice StockCode                         Description  Quantity  \
0  489434     21232      STRAWBERRY CERAMIC TRINKET BOX        24   
1  489434     21523  FANCY FONT HOME SWEET HOME DOORMAT        10   
2  489434     21871                 SAVE THE PLANET MUG        24   
3  489434     22041        RECORD FRAME 7" SINGLE SIZE         48   
4  489434     22064          PINK DOUGHNUT TRINKET POT         24   

                  InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00.000000   1.25      13085.0  United Kingdom  
1  2009-12-01 07:45:00.000000   5.95      13085.0  United Kingdom  
2  2009-12-01 07:45:00.000000   1.25      13085.0  United Kingdom  
3  2009-12-01 07:45:00.000000   2.10      13085.0  United Kingdom  
4  2009-12-01 07:45:00.000000   1.65      13085.0  United Kingdom

## EDA

### How many customers are repeat customers vs new?

### What are the most bought items?

### How often does a customer come back? Is there an average timeframe?